dynamicio/handlers/postgres.py

```diff
@@ -0,0 +1,163 @@
+
+    db_user: str
+    db_password: Optional[str]
+    db_host: str
+    db_port: int
+    db_name: str
+
+    @property
+    def connection_string(self) -> str:
+        """Build connection string out of components."""
+        password = f":{self.db_password}" if self.db_password else ""
+        return f"postgresql://{self.db_user}{password}@{self.db_host}:{self.db_port}/{self.db_name}"
+
+
+class ConfigurationError(Exception):
+    """Raised when configuration is wrong."""
+
+
+class PostgresResource(PostgresCredentials, BaseResource):
+    """...
+
+    Warning: Special case, where even if you explicitly don't validate, but give a pa_schema with strict="filter",
+    the generated sql query will only query specified columns.
+    """
+
+    table_name: Optional[str] = Field(None, description="SQL table name. Needs to be given if no sql_query is given")
+    sql_query: Optional[str] = Field(
+        None, description="SQL query. Will fetch schema defined columns if this is not given."
+    )
+    truncate_and_append: bool = False
+    kwargs: Dict[str, Any] = {}
+
+    def inject(self, **kwargs) -> "PostgresResource":
+        """Inject variables into stuff. Not in place."""
+        clone = deepcopy(self)
+        clone.db_user = inject(clone.db_user, **kwargs)
+        if db_password := self.db_password:
+            clone.db_password = inject(db_password, **kwargs)
+        clone.db_host = inject(clone.db_host, **kwargs)
+        clone.db_name = inject(clone.db_name, **kwargs)
+        if table_name := self.table_name:
+            clone.table_name = inject(table_name, **kwargs)
+        if sql_query := self.sql_query:
+            clone.sql_query = inject(sql_query, **kwargs)
+        return clone
+
+    def _check_injections(self) -> None:
+        """Check that all injections have been completed."""
+        check_injections(self.db_user)
+        if db_password := self.db_password:
+            check_injections(db_password)
+        check_injections(self.db_host)
+        check_injections(self.db_name)
+        if table_name := self.table_name:
+            check_injections(table_name)
+        if sql_query := self.sql_query:
+            check_injections(sql_query)
+
+    def _resource_read(self):
+        """Handles Read operations for Postgres."""
+        if not bool(self.sql_query) ^ bool(self.table_name):
+            raise ConfigurationError("PostgresResource must define EITHER sql_query OR table_name.")
+
+        if self.pa_schema is not None and (not self.sql_query and self.pa_schema.Config.strict):
+            # filtering can now be done at sql level
+            columns: List[str] = list(self.pa_schema.to_schema().columns.keys())  # type: ignore
+            table_to_query = table(self.table_name, *[column(name) for name in columns])
+            query = select(table_to_query.columns)
+            sql_query = str(query)
+        elif self.sql_query is None:
+            sql_query = f"SELECT * FROM {self.table_name}"
```

Comment:
Are the users supposed to pass in the `schema` too along with the table name?

e.g.

`public.cargo_movements`

Right now, the code suggests that I can pass in `table_name="cargo_movements"`, but there's no way to specify which schema it lives in

Reply:
I am not sure what you mean. If there is a pandera schema (defined when initialising the resource). Then it will use that schema to construct the query, rather than using `SELECT *`.

There is an issue though, that because of the current code structure this only works when initialising the resource with a schema, not when there is a schema passed at the time of calling `.read`.

Reply:
Ah, my bad. Yes we should add that functionality. Did this exist before?

Reply:
Actually, that doesn't seem to be the case. I think underneath it just assumes the `public` schema. Checking some postgres yamls in `distillation`, I can see that we just pass in the table name... I don't really understand that design choice. Seems to limit versatility